streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -263,26 +261,23 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
 
             long offset = globalConsumer.position(topicPartition);
             final Long highWatermark = highWatermarks.get(topicPartition);
-            final BatchingStateRestoreCallback stateRestoreAdapter =
-                (BatchingStateRestoreCallback) ((stateRestoreCallback instanceof
-                                                     BatchingStateRestoreCallback)
-                                                ? stateRestoreCallback
-                                                : new WrappedBatchingStateRestoreCallback(stateRestoreCallback));
+            final BatchingStateRestoreRecordCallback stateRestoreAdapter =
+                StateRestoreCallbackAdapter.adapt(stateRestoreCallback);
 
             stateRestoreListener.onRestoreStart(topicPartition, storeName, offset, highWatermark);
             long restoreCount = 0L;
 
             while (offset < highWatermark) {
                 try {
                     final ConsumerRecords<byte[], byte[]> records = globalConsumer.poll(pollTime);
-                    final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
+                    final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>();
                     for (final ConsumerRecord<byte[], byte[]> record : records) {
                         if (record.key() != null) {
-                            restoreRecords.add(KeyValue.pair(record.key(), record.value()));
+                            restoreRecords.add(record);
                         }
                     }
                     offset = globalConsumer.position(topicPartition);
-                    stateRestoreAdapter.restoreAll(restoreRecords);
+                    stateRestoreAdapter.restoreBatch(restoreRecords);
```

Comment:
The way it works is that now, we just go ahead and send the full consumer records, and the adapter will take care of converting it to `KeyValue`s if that's what the callback needs.